### PR TITLE
JES Auth rework for Firecloud

### PR DESCRIPTION
**Pull request by Horneth**
This reworks the JES Authentication schema in such a way that:

- All JES calls (launching a VM) are requested using the Cromwell Service Account (CSA) passed in through the configuration.
- The upload / delete of the authentication file to GCS is also done using CSA.
- All other GCS interactions are done using the user's credentials (with refresh token coming from WF options + clientID/secrets in the conf).
- a `billing_bucket` workflow option has been added that sets the gcs path where cromwell will write the auth.json file

### CODE REVIEW

File: src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala [hunk 1]
```diff
@@ -31,6 +32,7 @@ object JesAttributes {
     "applicationName",
     "project",
     "baseExecutionBucket",
+    "scratchBucket",
```
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala]: Hmm baseExecutionBucket is where everything happens (call directories, logs, outputs, exec.sh...). What this PR does is use user's crendentials (if available) for all interactions with this bucket.
The scratchBucket still uses Cromwell credentials (service account) and is only used to upload/delete the auth file and users shouldn't have access to it.
So both are needed I think.

File: src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala [hunk 2]
```diff
@@ -20,7 +35,12 @@ object ProductionJesConnection {
  */
 trait JesAuthentication { self: JesConnection =>
 
-  def authenticated[A](f: JesInterface => A) = f(jesConnection)
+  def authenticatedAsCromwell[A](f: JesInterface => A) = f(jesCromwellConnection)
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: Why was this renamed to `authenticatedAsCromwell`?

Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: because there is an `authenticatedAsUser` now so I thought it was clearer to change the old `authenticated` so you always know as what you're authenticated.

File: src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala (hunk 1 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala]: Discussed in person

File: src/main/scala/cromwell/engine/backend/jes/JesBackendCall.scala [hunk 3]
```diff
@@ -58,10 +58,13 @@ class JesBackendCall(val backend: JesBackend,
...
   lazy val cmdInput = JesInput(ExecParamName, gcsExecPath.toString, localFilePathFromRelativePath(JesExecScript))
   lazy val diskInput = JesInput(WorkingDiskParamName, LocalWorkingDiskValue, Paths.get(JesCromwellRoot))
-  
+  lazy val refreshToken = workflowDescriptor.workflowOptions.get(JesBackend.RefreshTokenOptionKey).toOption
+
   def standardParameters = Seq(stderrJesOutput, stdoutJesOutput, rcJesOutput, diskInput)
 
-  def downloadRcFile = authenticated { connection => GoogleCloudStoragePath.parse(callGcsPath + "/" + RcFilename).map(connection.storage.slurpFile) }
+  def downloadRcFile = authenticatedAsUser(this) { storage =>
+    GoogleCloudStoragePath.parse(callGcsPath + "/" + RcFilename).map(storage.slurpFile)
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/jes/JesBackendCall.scala]: Why not use string interpolation here?

File: src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala (hunk 2 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: I think I'd prefer something like `authenticateAsCromwell` and `authenticateAsUser` - the ordering sounds more natural to me. OTOH I always did terribly in grammar class and such when i was young

Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: Oh I see, I missed the `authenticatedAsUser` piece.

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 4]
```diff
@@ -571,8 +583,9 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
   override def backendType = BackendType.JES
 
   def gcsAuthFilePath(descriptor: WorkflowDescriptor): String = {
-    val wfPath = workflowGcsPath(descriptor)
-    s"$wfPath/auth.json"
+    val bucket = jesConf.scratchBucket
+    val fileName = s"${descriptor.id}_auth.json"
+    s"$bucket/$fileName"
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: maybe just `s"${jesConf.scratchBucket}/${descriptor.id}_auth.json"`?

File: src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala (hunk 1 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesAttributes.scala]: Might be good to immortalize this in comments.  :smile:

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 5]
```diff
@@ -62,6 +61,18 @@ object JesBackend {
   // Decoration around WorkflowDescriptor to generate bucket names and the like
   implicit class JesWorkflowDescriptor(val descriptor: WorkflowDescriptor) extends JesBackend {
     def callDir(key: CallKey) = callGcsPath(descriptor, key.scope.name, key.index)
+
+   /*
+   * User storage is available if googleSecrets configuration has been set and q refresh token has been passed in the workflow options.
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: funky comment formatting, "q refresh token"?

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 6]
```diff
@@ -329,20 +340,21 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
     val fileContent =
       s"""
          |#!/bin/bash
+         |export _JAVA_OPTIONS=-Djava.io.tmpdir=$JesCromwellRoot/tmp
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: this doesn't seem obviously related to the purpose of the PR

Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: This is part of another PR

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 7]
```diff
@@ -208,7 +219,7 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
    * First queries for the existence of the auth file, then deletes it if it exists.
    * If either of these operations fails, then a Future.failure is returned
    */
-  override def cleanUpForWorkflow(workflow: WorkflowDescriptor)(implicit ec: ExecutionContext): Future[Unit] = authenticated { connection =>
+  override def cleanUpForWorkflow(workflow: WorkflowDescriptor)(implicit ec: ExecutionContext): Future[Unit] = authenticatedAsCromwell { connection =>
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: This method could maybe have a better name, it almost sounds like it's doing compaction rather than deleting the auth file.

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 8]
```diff
@@ -329,20 +340,21 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
...
-    def attemptToUploadObject(priorAttempt: Option[Unit]) = authenticated { _.storage.uploadObject(backendCall.gcsExecPath, fileContent) }
+    def attemptToUploadObject(priorAttempt: Option[Unit]) = authenticatedAsUser(backendCall) { _.uploadObject(backendCall.gcsExecPath, fileContent) }
 
     val log = workflowLogger(backendCall.workflowDescriptor)
     withRetry(attemptToUploadObject, log, s"${workflowLoggerWithCall(backendCall).tag} Exception occurred while uploading script to ${backendCall.gcsExecPath}")
   }
 
   private def createJesRun(backendCall: BackendCall, jesParameters: Seq[JesParameter], runIdForResumption: Option[String]): Try[Run] =
-    authenticated { connection =>
+    authenticatedAsCromwell { connection =>
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Perhaps it's just a symptom of my ignorance of our current scheme, but it feels strange we're creating the run as Cromwell and then doing everything inside as the user...

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: That is indeed our current scheme. To the cloud!

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 6 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Right I keep adding this one back when I need to start GotC jobs but sometimes forget to remove it  It would be nice if we could get it merged in btw as the single sample won't run without it for large files.

File: src/main/scala/cromwell/binding/IOInterface.scala [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package cromwell.binding
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/IOInterface.scala]: Why is this in the binding package?

Comment (Horneth) [inline src/main/scala/cromwell/binding/IOInterface.scala]: Hmm it could be in engine too I guess yes, it made sense to me to put it in binding because it's supposed to give the Wdl Library function (which are in bindings) a way to do IO, and the implementation would be backend dependant. But I don't mind moving it to engine.

Comment (geoffjentry) [inline src/main/scala/cromwell/binding/IOInterface.scala]: I think we need to start looking at the binding package as "if we were to construct a workflow out of scala objects from an arbitrary workflow language and for an arbitrary backend, what would we have in there?" - it's not there now but looking ahead it'll make life less painful to try to apply that reasoning going forward.

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 10]
```diff
@@ -239,7 +245,7 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
     new JesBackendCall(this, workflowDescriptor, key, locallyQualifiedInputs, abortRegistrationFunction)
   }
 
-  override def engineFunctions: WdlStandardLibraryFunctions = new JesEngineFunctionsWithoutCallContext()
+  override def engineFunctions(interface: IOInterface): WdlStandardLibraryFunctions = new JesEngineFunctionsWithoutCallContext(interface)
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: if I'm reading this right it means that technically one could have an interface being passed in for the incorrect backend?

Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: I think yes :s, the whole problem here is that we need to be able to read files when we evaluate expressions in the workflow declaration, which happens when the workflow descriptor is created.
This means that the workflow descriptor needs something to pass to the engineFunctions so they can read files (for `read_lines` etc...). However we don't want to introduce backend specific terminology/logic at the workflow descriptor level, which is why IOInterface is completely generic.
Maybe with some kind of [T] parameter it would be possible to make sure that we pass the right interface.

File: src/main/scala/cromwell/binding/IOInterface.scala (hunk 9 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/IOInterface.scala]: Well then I guess it depends on do we consider that any workflow language would need basic IO capabilities (read a file, write a file, list files in a directory...), or not. It's not obvious I don't know if CWL would actually have the need for this in the language itself.

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 10 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Added a `type IOInterface` as per @geoffjentry suggestion that is overriden by the backend implementation (similarly to BackendCall), so that passing the wrong IOInterface implementation won't compile.

File: src/main/scala/cromwell/binding/IOInterface.scala (hunk 9 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/IOInterface.scala]: When I was looking at it earlier I was hung up on that abstract type bit we discussed in person. I think you're right here.

File: src/main/scala/cromwell/engine/backend/local/SharedFileSystem.scala [hunk 11]
```diff
@@ -72,11 +72,37 @@ object SharedFileSystem {
   }
 }
 
+class SFSIOInterface extends IOInterface {
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/local/SharedFileSystem.scala]: This could probably use a better name... I think `SharedFileSystemIOInterface` is better than the abbreviated version

File: src/main/scala/cromwell/util/google/GoogleCloudStorage.scala [hunk 12]
```diff
@@ -24,29 +26,62 @@ object GoogleCloudStorage {
...
-  def listBucket(bucketName: String): GcsBucketInfo = {
-    val getBucket = client.buckets().get(bucketName)
-    getBucket.setProjection("full")
-    val bucket: Bucket = getBucket.execute()
+case class GoogleCloudStorage(client: Storage) extends IOInterface {
 
-    new GcsBucketInfo(bucketName, bucket.getLocation, bucket.getTimeCreated, bucket.getOwner)
+  import GoogleCloudStoragePath._
+
+  /* **** IO Interface **** */
```
Comment (scottfrazer) [inline src/main/scala/cromwell/util/google/GoogleCloudStorage.scala]: This comment doesn't really add much

File: src/main/scala/cromwell/util/google/GoogleCloudStorage.scala [hunk 13]
```diff
@@ -24,29 +26,62 @@ object GoogleCloudStorage {
...
+  def writeTempFile(path: String, prefix: String, suffix: String, content: String): String = {
+    val fullPath = s"$path/$prefix${UUID.randomUUID()}$suffix"
+    writeFile(fullPath, content)
+    path
+  }
+
+  //TODO: improve to honor pattern ?
+  def glob(path: String, pattern: String): Seq[String] = listContents(path).toSeq
+
+  /* ************************ */
```
Comment (scottfrazer) [inline src/main/scala/cromwell/util/google/GoogleCloudStorage.scala]: Same with this comment

**Comment by scottfrazer** [comment]
:+1:

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 14]
```diff
@@ -208,24 +212,29 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
...
-      case Failure(ex) =>
-        log.error(s"Could not query for the existence of the auth file $gcsAuthFile", ex)
-        Future.failed(ex)
-      case _ => Future.successful(Unit)
-    }
+  override def cleanUpForWorkflow(workflow: WorkflowDescriptor)(implicit ec: ExecutionContext): Future[Unit] = {
+    Future(gcsAuthFilePath(workflow)) map { path =>
+      deleteAuthFile(path, workflowLogger(workflow))
+      ()
+    } recover { case _ => } // No need to fail here, it just means that we didn't have an auth file in the first place so no need to delete it.
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: That's one reason the `Future` might have failed, but potentially there might be other less benign reasons.  Could this be more selective on the `Throwable` type?

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 10 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Not that this needs to be addressed now, but we've discussed having the local backend be able to work with GCS.

File: src/main/scala/cromwell/util/google/GoogleCloudStorage.scala [hunk 15]
```diff
@@ -24,29 +26,58 @@ object GoogleCloudStorage {
...
+  def readFile(path: String): String = {
+    new String(downloadObject(path), "UTF-8")
   }
 
-  def listContents(gcsPath: String): Iterable[String] = listContents(GoogleCloudStoragePath(gcsPath))
+  def exists(path: String): Boolean = {
+    val getObject = client.objects.get(path.bucket, path.objectName)
+    Try(getObject.execute) match {
+      case Success(_) => true
+      case Failure(ex: GoogleJsonResponseException) if ex.getStatusCode == 404 => false
```
Comment (mcovarr) [inline src/main/scala/cromwell/util/google/GoogleCloudStorage.scala]: maybe this should be retryable?

File: src/main/scala/cromwell/engine/backend/Backend.scala [hunk 16]
```diff
@@ -77,7 +78,12 @@ trait Backend {
   /**
    * Engine functions that don't need a Call context (e.g. read_lines(), read_float(), etc)
    */
-  def engineFunctions: WdlStandardLibraryFunctions
+  def engineFunctions(interface: IOInterface): WdlStandardLibraryFunctions
+
+  /**
+    * Interface to be used primarily by engine functions requiring IO capabilities.
+    */
+  def IOInterface(workflowOptions: WorkflowOptions): IOInterface
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: strange method name, it makes it look like an apply method or case class constructor is being invoked

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 17]
```diff
@@ -571,8 +580,11 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
   override def backendType = BackendType.JES
 
   def gcsAuthFilePath(descriptor: WorkflowDescriptor): String = {
-    val wfPath = workflowGcsPath(descriptor)
-    s"$wfPath/auth.json"
+    // If we are going to upload an auth file we need a valid GCS path passed via workflow options.
+    val bucket = descriptor.workflowOptions.get(AuthFilePathOptionKey) getOrElse {
+      throw new Throwable(s"$AuthFilePathOptionKey has not been found in workflow options.")
+    }
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: should this be failing faster?

Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: It actually still fails relatively early, because the first time we call this is when trying to write the auth file which is during `initializeForWorkflow`.

File: src/main/scala/cromwell/engine/backend/Backend.scala (hunk 16 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: Yeah I'll add a `get` in front of it

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: Get is unidiomatic scala

File: src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala [hunk 18]
```diff
@@ -1,18 +1,32 @@
...
 
 /**
  * Trait for JesConnection
  */
 trait JesConnection {
-  def jesConnection: JesInterface
+  def jesCromwellConnection: JesInterface
+
+  /**
+    * This method should try its best to provide a GCS connection setup with the user's crendentials.
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: crendentials => credentials

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 17 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: oh nm, this is per workflow reading from workflow options, I thought it was config...

File: src/main/scala/cromwell/engine/backend/Backend.scala (hunk 16 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: what about `monadicLambdaFunctorIOInterfaceWhichHasNothingToDoWithJava` then ? :smile:

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 19]
```diff
@@ -600,4 +612,16 @@ class JesBackend extends Backend with LazyLogging with ProductionJesAuthenticati
   override def findResumableExecutions(id: WorkflowId)(implicit ec: ExecutionContext): Future[Map[ExecutionDatabaseKey, JobKey]] = {
     globalDataAccess.findResumableJesExecutions(id)
   }
+
+  /**
+    * Will try to instantiate a user authenticated IOInterface if possible.
+    * If not, will default back to the Cromwell authenticated IOInterface.
+    */
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: Clearly I'm not completely up to date on our current auth scheme, why is this desirable behavior?

File: src/main/scala/cromwell/engine/backend/Backend.scala (hunk 16 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: Now that I'm looking at this on a screen larger than an iPhone, I actually disagree w/ @mcovarr although I'd probably call it `ioInterface`

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: That's not to say that your new suggestion isn't awesome

Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/Backend.scala]: maybe if it's in Greek characters

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala (hunk 19 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: So in the current state you can provide `clientId/Secrets` in the configuration.
If you do, we assume that you want to access GCS as a User for everything except the write of the auth file (which has to be done by cromwell SA). You will then have to provide a bunch of parameters in the workflow options (refresh_token, bucket for the auth file). That's how cromwell is going to run in a firecloud context.
But for GITC there's no user authentication, everything is done with the service account.
This is allows the backend to support both modes, by providing a user authenticated interface if we have everything to do it, and if not then the normal Service Account is used.

File: src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala [hunk 20]
```diff
@@ -20,7 +34,12 @@ object ProductionJesConnection {
  */
 trait JesAuthentication { self: JesConnection =>
 
-  def authenticated[A](f: JesInterface => A) = f(jesConnection)
+  def authenticateAsCromwell[A](f: JesInterface => A) = f(jesCromwellConnection)
+
+  /**
+   * Important note: Will default back to cromwell authentication if the configuration for user authentication has not been set or if the refreshToken has been supplied.
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: Is this saying that if a workflow is submitted with a refresh token the customer's billing project is billed, otherwise we bill ourselves?

Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: No because `authenticateAsUser` only gives you back a `GoogleCloudStorage` object, so you can only read/write stuff into GCS with it, not request compute.
So this only allows you to read/write stuff as a user if they give a refresh token, or as cromwell if they don't.

Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: Also it's actually the project you request compute in that determines who is getting billed. So for us to get billed instead of them we would have to receive our google project in the workflow options instead of the user's.

File: src/main/scala/cromwell/util/google/GoogleCloudStorage.scala [hunk 21]
```diff
@@ -24,29 +26,58 @@ object GoogleCloudStorage {
   case class GcsBucketInfo(bucketName: String, location: String, timeCreated: DateTime, owner: Owner)
 }
 
+trait GCSIOInterface extends IOInterface {
+
+}
```
Comment (mcovarr) [inline src/main/scala/cromwell/util/google/GoogleCloudStorage.scala]: this appears to be unused

File: src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala (hunk 20 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/authentication/JesAuthentication.scala]: Discussed further in person, we'll want to bring back a more binary choice of Cromwell operational modes: user auth vs service account.

**Comment by mcovarr** [comment]
:+1: looks good overall delta a few comments
